docs/src/examples/gradient_based_optimization.md

```diff
@@ -0,0 +1,180 @@
+using Ipopt
+using PythonPlot
+using Test
+include(__TASOPTindices__)
+```
+
+Initialize arrays used for plotting as well as the finite difference parameter
+
+```julia
+design_variables = []
+objective_array = []
+
+epsilon = 1e-5
+```
+
+Load the default `aircraft` model and size it to get initial values:
+
+```julia
+default_model = load_default_model()
+size_aircraft!(default_model)
+```
+
+## Setting Optimization Parameters
+
+### Set the input parameters
+
+```julia
+input_params = [
+    :(ac.parg[igAR]), 
+    :(ac.para[iaCL,ipcruise1:ipcruise2,1]),
+    :(ac.parg[igsweep]),
+    :(ac.pare[ieTt4, ipcruise1:ipcruise2, 1]),
+    :(ac.pare[iepif, ipcruise1, 1]) ,
+]
+
+# Formatting params for usage 
+params = map(p -> TASOPT.format_params(TASOPT.expr_to_string(p)), input_params)
+```
+
+### Set the Upper and Lower limits for all design variables as well as initial values
+
+```julia
+lower      = [9.0 , 0.53, 25.0, 1400.0, 1.25]
+upper      = [11.0, 0.60, 30.0, 1650.0, 2.0 ] 
+initial    = [10.5, 0.57, 26.0, 1580.0, 1.685]
+
+```
+
+## Objective Function
+
+```julia
+function size_ac(x...)
+    # println("State: ",x)
+    ac = deepcopy(default_model)
+    # Set params
+    for (i,x_i) in enumerate(x)
+        field_path, index = params[i]
+        TASOPT.setNestedProp!(ac, field_path, x_i, index)
+    end
+    # Size aircraft
+    push!(design_variables, x)
+    try
+        size_aircraft!(ac,printiter=false)
+        push!(objective_array, ac.parm[imPFEI])
+        return ac.parm[imPFEI]
+    catch
+        println("wsize FAILED")
+        push!(objective_array, NaN)
+        return 1.0e12
+    end
+end
+```
+
+## Gradient Function
+
+```julia
+function fdiff_all!(g, x...)
+    # Reset model
+    ac = deepcopy(default_model)
+    # Set current state
+    for (i,x_i) in enumerate(x)
+        field_path, index = params[i]
+        TASOPT.setNestedProp!(ac, field_path, x_i, index)
+    end
+    # Calculate gradient
+    if (size(g)[1] >0)
+        gradients = TASOPT.get_sensitivity(input_params; model_state=ac, eps=epsilon, optimizer=true)
+        for (k,grads) in enumerate(gradients)
+            g[k] = grads
+        end
+    end
+end
+```
+
+## Running the optimization
+
+```julia
+function exampleOPT()
+    model = Model(Ipopt.Optimizer)
+
+    # For almost localy solved (Takes much shorter time, almost solves):
+    # set_optimizer_attribute(model, "acceptable_tol", 1e-2)
+    # set_optimizer_attribute(model, "acceptable_iter", 1)
+    
+    @variable(model, lower[i] <= x[i=1:size(initial)[1]] <= upper[i], start=initial[i])
+
+    function eval_f(x...)
+        return size_ac(x...)
+    end
+
+    register(model, :eval_f, size(initial)[1], eval_f, fdiff_all!; autodiff = false)
+
+    @NLobjective(model, Min, eval_f(x...))
+
+    optimize!(model)
+    println("""
+    termination_status = $(termination_status(model))
+    primal_status      = $(primal_status(model))
+    objective_value    = $(objective_value(model))
+    """)
+
```

Comment:
I think it would be valuable to add constraints too. In my experience with IpOpt, setting up the constraints was harder than any other thing. I believe the syntax is of the type:
`@NLexpressions(model, begin
    q_1, size_ac(vars...)
    q_2, FuelVolCstr()
    q_3, SpanCstr()
    q_4, ClimbCstr()
    q_5, Tt3Cstr()
    q_6, TmetalCstr()
    q_7, MTOWCstr()
    q_8, FanDiamCstr()
    q_9, FanSizeCstr()
    q_10, FanWingCstr()
    end)`